### PR TITLE
saml clock skew test

### PR DESCRIPTION
**Pull request by vramik**

### CODE REVIEW

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLClockSkewAdapterTest.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+    private static final String CONTEXT_ROOT = "sales-post-clock-skew";
+    private static final String DEPLOYMENT_NAME_1 = CONTEXT_ROOT + "_1";
+    private static final String DEPLOYMENT_NAME_2 = CONTEXT_ROOT + "_2";
+
+    @ArquillianResource private Deployer deployer;
+
+    @Deployment(name = DEPLOYMENT_NAME_1, managed = false)
+    protected static WebArchive salesPostClockSkewServlet1() {
+        return samlServletDeployment(CONTEXT_ROOT, DEPLOYMENT_NAME_1, CONTEXT_ROOT + "/WEB-INF/web.xml", 3, AdapterActionsFilter.class, PublicKeyLocator.class, SendUsernameServlet.class);
```
Comment (vramik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLClockSkewAdapterTest.java]: @steevebtib 
Currently there are two deployments, both of them are the same war except the value (in seconds) for clock skew. Both deployments are managed manually, before test the deployment is deployed and after test is undeployed. 

Feel free to add other similar deployments for more tests.

**Comment by vramik** [comment]
@steevebtib pls take a look at the tests and let me know if there is anything unclear. Thank you.

File: saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/XMLTimeUtil.java [hunk 2]
```diff
@@ -102,15 +102,19 @@ public static XMLGregorianCalendar getIssueInstant(String timezone) {
         GregorianCalendar gc = new GregorianCalendar(tz);
         XMLGregorianCalendar xgc = dtf.newXMLGregorianCalendar(gc);
 
+        Long offsetMilis = TimeUnit.MILLISECONDS.convert(Time.getOffset(), TimeUnit.SECONDS);
+        if (offsetMilis != 0) {
+            logger.debug(XMLTimeUtil.class.getName() + " timeOffset: " + offsetMilis);
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/XMLTimeUtil.java]: Please use `logger.debugf` instead (dtto below)

File: saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/XMLTimeUtil.java [hunk 3]
```diff
@@ -179,12 +183,10 @@ public static boolean isValid(XMLGregorianCalendar now, XMLGregorianCalendar not
...
      *
      * @return
-     *
-     * @throws org.keycloak.saml.common.exceptions.ParsingException
      */
-    public static Duration parseAsDuration(String timeValue) throws ParsingException {
+    public static Duration parseAsDuration(String timeValue) {
         if (timeValue == null) {
-            PicketLinkLoggerFactory.getLogger().nullArgumentError("duration time");
+            throw logger.nullArgumentError("duration time");
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/XMLTimeUtil.java]: -1 revert

File: saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/XMLTimeUtil.java [hunk 4]
```diff
@@ -210,7 +212,7 @@ public static Duration parseAsDuration(String timeValue) throws ParsingException
      *
      * @throws ParsingException
      */
-    public static XMLGregorianCalendar parse(String timeString) throws ParsingException {
+    public static XMLGregorianCalendar parse(String timeString) {
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/XMLTimeUtil.java]: -1, revert

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLClockSkewAdapterTest.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+//@AppServerContainer(ContainerConstants.APP_SERVER_TOMCAT7)
+//@AppServerContainer(ContainerConstants.APP_SERVER_TOMCAT8)
+//@AppServerContainer(ContainerConstants.APP_SERVER_TOMCAT9)
+//@AppServerContainer(ContainerConstants.APP_SERVER_JETTY92)
+//@AppServerContainer(ContainerConstants.APP_SERVER_JETTY93)
+//@AppServerContainer(ContainerConstants.APP_SERVER_JETTY94)
+public class SAMLClockSkewAdapterTest extends AbstractServletsAdapterTest {
+
+    private static final String CONTEXT_ROOT = "sales-post-clock-skew";
+    private static final String DEPLOYMENT_NAME_1 = CONTEXT_ROOT + "_1";
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLClockSkewAdapterTest.java]: Could the names (both the constants and method names) reflect the actual clock skew (e.g. `DEPLOYMENT_NAME_3_SEC`, `DEPLOYMENT_NAME_30_SEC`)?

File: saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/XMLTimeUtil.java (hunk 2 above)
Comment (vramik) [reply saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/XMLTimeUtil.java]: There is no such method in org.keycloak.saml.common.PicketLinkLogger. We would either add the method there or leave it as it is. I'd vote for latter option.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLClockSkewAdapterTest.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+@AppServerContainer(ContainerConstants.APP_SERVER_WILDFLY_DEPRECATED)
+@AppServerContainer(ContainerConstants.APP_SERVER_EAP)
+@AppServerContainer(ContainerConstants.APP_SERVER_EAP6)
+@AppServerContainer(ContainerConstants.APP_SERVER_EAP71)
+//@AppServerContainer(ContainerConstants.APP_SERVER_TOMCAT7)
+//@AppServerContainer(ContainerConstants.APP_SERVER_TOMCAT8)
+//@AppServerContainer(ContainerConstants.APP_SERVER_TOMCAT9)
+//@AppServerContainer(ContainerConstants.APP_SERVER_JETTY92)
+//@AppServerContainer(ContainerConstants.APP_SERVER_JETTY93)
+//@AppServerContainer(ContainerConstants.APP_SERVER_JETTY94)
```
Comment (vramik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLClockSkewAdapterTest.java:59]: @steevebtib pls note, I haven't tried either tomcat or jetty.

File: saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/XMLTimeUtil.java (hunk 2 above)
Comment (vramik) [reply saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/XMLTimeUtil.java]: As discussed offline, `if (logger.isDebugEnabled())` is used

File: saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/XMLTimeUtil.java (hunk 3 above)
Comment (vramik) [reply saml-core/src/main/java/org/keycloak/saml/processing/core/saml/v2/util/XMLTimeUtil.java]: reverted, thanks for clarification

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLClockSkewAdapterTest.java (hunk 5 above)
Comment (vramik) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/adapter/servlet/SAMLClockSkewAdapterTest.java]: good idea, incorporated

**Comment by steevebtib** [comment]
@vramik Thank you for your example, I am going to implement it today, is it possible to integrate this pull request to my commit ?

**Comment by vramik** [comment]
@steevebtib yeah, go ahead, it was meant to be merged into your fork, thank you for your time spent on in.

**Comment by steevebtib** [comment]
@vramik It is done, when I finish should I merge your commit and mine and redo the force push?

**Comment by vramik** [comment]
@steevebtib yeah, once you're done, quash commits into one and then do force push.

**Comment by steevebtib** [comment]
@vramik Here is what I have done up to now, I have tested against arquilian containers inside my IDE (IntelliJ) for Tomcat 7, 8, Jetty 9.2, 9.3, 9.4 with Arquillian runners. I cannot find an Arquillian container for Tomcat 9, could you please explain me where to find it ? Before pushing the update, I will do a mvn clean install on the whole project to be sure I did not miss anything.

**Comment by vramik** [comment]
@steevebtib there is module for [tomcat9](https://github.com/keycloak/keycloak/tree/master/testsuite/integration-arquillian/servers/app-server/tomcat/tomcat9), there is info how to run it: https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/HOW-TO-RUN.md#tomcat

**Comment by steevebtib** [comment]
@vramik thanks for the feedback

**Comment by steevebtib** [comment]
@vramik I uncommented the missing containers and ran the tests and I got no errors, for me that's it or did I miss something ?

**Comment by vramik** [comment]
@steevebtib if it works then it's great.

**Pull request merged**